### DISCUSSION TITLE
Bluetooth version

### DISCUSSION THREAD

**Discussion by ams1** [hrvach/deskhop#100] [Ideas]
Hi,

Long story short: a while back I had the **idea** to do a ~similar project, but given _**execution is key**_ and given I lacked c/c++ and hardware skills... I failed 😄 - true, I learned a LOT while trying really hard to implement it.
And now I stumbled on this awesome repo - THANK you for this!

Given I saw that bluetooth is on the deskhop roadmap - and it was one of my main concerns - in the below picture I tried to adapt what I had in mind to the deskhop context - hopefully our visions match.

![image](https://github.com/hrvach/deskhop/assets/10154357/e4be86fb-8b87-4c45-a43a-cd5e24903836)

(1) BLE Peripheral, ex: 
  * [https://github.com/bluekitchen/btstack/tree/master/example/hog_keyboard_demo.c](https://github.com/bluekitchen/btstack/tree/master/example/hog_keyboard_demo.c)

(2) BLE Central, ex: 
  * [https://github.com/ricardoquesada/bluepad32](https://github.com/ricardoquesada/bluepad32)
  * [https://github.com/bluekitchen/btstack/tree/master/example/hog_host_demo.c](https://github.com/bluekitchen/btstack/tree/master/example/hog_host_demo.c)

### So, IF we decide to attach 2 x Pico **_W_** modules on the extension board (instead of 2 x 'normal' Pico):
  * PROs:
    * Pico W is not so expensive and awesome
    * we could use one of the Pico W to connect to other BLE peripherals - as I've figured the [bluepad32](https://github.com/ricardoquesada/bluepad32) lib takes care of this - allows up to 4 devices (I'll test this soon) - and be able to use BT mice, keyboards, playstation controllers (awesome)
    * we could use the other Pico W to emulate a keyboard and mouse - computers connect to this Bluetooth Low Energy endpoint
    * or we could use both as BLE keyboards (as I found, when in BLE Peripheral mode, only 1 central can connect - by BLE standard design)
  * CONs:
    * given the Pico W is FCC certified using the onboard antenna, I think there is NO future option to order ex from JLCPCB a board with the same Bluetooth Chip already soldered (like I saw you have planned for the non-W option) - I think this is why there aren't any Pico W clones around => the Pico W will always have to be ordered separately and attached (I think)
    * according to the Pico [datasheet](https://datasheets.raspberrypi.com/picow/pico-w-datasheet.pdf) chapter 3.8, the area on the PCB below (or above) the Pico W wireless chip cannot have any metal - so no traces can pass there (I saw a design where the PCB had a physical rectangle cutout there) - so some changes would have to be made to the board (again that's my non-expert opinion)
    * there might be some issues when using PIO USB Host + Bluethooth + communication between the boards - I saw a design somewhere that additionally added a dedicated USB Host chip (not sure if because of this)
    * I think the current task-to-cores architecture would need to be adapted - ex. saw bluepad32 consumes a core

As a non-expert, it took me A WHILE 😅 to find all the pieces of this puzzle, I haven't yet assembled all of them (not even sure if it's possible), but if you decide to go in this direction COUNT ME IN (testing, some money etc.)!

With best regards,
Marius

**Comment by hrvach** [comment]
<img width="1150" alt="Screenshot 2024-06-16 at 18 52 56" src="https://github.com/hrvach/deskhop/assets/7192749/b164bf75-942a-4cc0-ab38-2736b2c8f821">
This is my original bluetooth concept, the idea being that it's cheaper and simpler than the original but it's a bit paused for now until I finish some of the features planned for the "classic" box. What I've found a bit annoying is the slow polling speed, much lower than 1000 hz mouse I hooked up to it, so I have to figure out how to approach this... I'll most likely keep working on it, but the existing users have priority. And yes, the antenna location is completely ignored while in development, at a later stage RF will be taken into account and pcb adapted.

**Comment by ams1** [comment]
@hrvach thank you so much for sharing the concept - it's basically my geeky Christmas wish. 🎄 

Regarding the polling speed, indeed also my experiments with bluetooth on the Pico W  (not via USB Host - I didn't manage to get PIO USB Host running alongside BLE...) showed that when I go above ~100 - 150 events per second it starts to lag.

Besides throttling via software (ex. aggregating/skipping events...), one "solution" might be to use a cheaper mouse that doesn't send so many events per second - a sort of natural throttling 😄. Probably many people don't use 1000Hz mice.

Speaking of many people - that also cannot solder 😁 -, would you 🙏 consider adding to the board the necessary footprint for the [Clipzin connectors](https://www.farnell.com/datasheets/3775376.pdf) **that would allow the Pico W to be snapped in instead of requiring soldering**?

![image](https://github.com/hrvach/deskhop/assets/10154357/65586cda-38b4-4af4-8637-80f4f0533350)

* CONs:
  * I know Clipzins are [expensive ](https://de.farnell.com/en-DE/edac/299-020-299198-1/card-edge-conn-single-20pos-solder/dp/4069520) (at least when bought in small qty) 
  * and that they cause the board to be a bit wider, BUT
* PROs:
  * Clipzins allow 2.7 mm of headroom between the Pico W and your board - you could use that space underneath the Pico W to maybe place other ICs like... an additional RP2040... for... I don't know, I'm just trying to "sell" you the Clipzin thing because many ppl - me included - don't solder 😅
  * and Clipzins would require NO SOLDERING on behalf of the user.
  * Also maybe quickly changing the Pico W boards... with different firmware...

I once wanted to offer a geeky Pico-W Christmas present to some of my friends and gave up because of the soldering.

I would love to order this board from here/somewhere, order the FCC-certified Pico W from somewhere else (ideally the same place) and just click them together...

Again thank you for sharing this - it's the first picture I see for Pio USB Host + Pico W (something I've been searching for a loong time)... It's beautiful!

**Reply by hrvach** [reply]
I got virtually identical results - after 100-150 events/sec it starts lagging. This could probably be handled in software and it would be simpler and slightly more affordable than the wired version. I'll probably move on to work on that after I polish the "wired" one. My only pet peeve is using wireless for keyboards/mice because of vulnerabilities getting discovered in popular protocols and then never getting fixed, but having a single board with literally 2-3 components slapped on is a very interesting proposal.

That Clipzin concept is pretty cool, I must say! Also, perhaps try soldering? It's a fun activity, you get hooked quite easily. :-)
